Additional support linker tests

tlakollo opened the pull request:
Rename ILLink.Tests* to Mono.Linker.Tests*
Add editorconfig files to avoid using runtime formating for ILLink.Shared and Mono.Linker.Tests* projects
Add tests project into CI subset clr.toolstests
Add unchecked clause to silence an error while building
Add ProducedBy.NativeAot and only verify for ProducedBy.NativeAot in ResultChecker
Add RequiresCapability tests
CheckAndReportRequires now uses a diagnosticContext allowing to make the method static
Call CheckAndReportRequires from the ReflectionMarker to support diagnostics in reflection calls
Add additional diagnosticContext depending on the attribute they relate to

Commit:
Rename ILLink.Tests* or Mono.Linker.Tests*

Commit:
Add test project into CI subset clr.toolstests

Commit:
Add unchecked clause to silence an error while building
Add ProduceBy.NativeAOT and verify only verify ProducedBy.NativeAOT in
ResultChecker

Commit:
Add RequiresCapability tests

vitek-karas commented:
Not a fan of the reformatting. There will be a time period (who knows how long) when we will have the tests in both linker and runtime repo. Migrating changes between the two should be trivial (ideally just copy over new version of the files), but with the formatting changes it will be a hassle.

If we need to tweak some editorconfigs or something to make this work in the runtime repo I think it's worth it.

There will be a set of files (everything in the Shared project and almost everything int he Test projects) which should be kept as exact copies from the linker repo. If we need to make a change in those files the change should be applied to both repos (eventually). For this reason the files should be kept exactly the same.

Could you please somehow revert the formatting changes (for the test projects this can be done by copying the files again from linker repo)?

Commit:
Move CheckAndReportRequires to use a diagnosticContext allowing to make
the method static
Call CheckAndReportRequires from the ReflectionMarker to support
diagnostics in reflection calls
Add additional diagnosticContext depending on the attribute they relate
to

Commit:
Add editorconfig files with the linker format on them
Reformat files with mixing format to linker format

Commit:
Merge branch 'SyncToLinkerAndTests' of https://github.com/vitek-karas/runtime into AdditionalSupportLinkerTests

Commit:
Use a single editorconfig since they will conflict when importing
projects
